### ISSUE TITLE
pcolormesh: shading='nearest' and non-monotonic coordinates

### ISSUE THREAD

**Issue by mathause** [matplotlib/matplotlib#18317]
# Bug report

**Bug summary**

#16258 deprecates `shading="flat"` when data and coordinates have the same shape and introduces `shading="nearest"` which interpolates the coordinates. However, the logic to interpolate the coordinates breaks down if they are not monotonically increasing.

https://github.com/matplotlib/matplotlib/blob/85117715a154d1027c6e204d2ee581fa787f423d/lib/matplotlib/axes/_axes.py#L5684

Of course non-monotonic coordinates are a bit exotic but they make sense for geographic data.

**Code for reproduction**

```python
import numpy as np

def _interp_grid(X):
    # helper for below
    if np.shape(X)[1] > 1:
        dX = np.diff(X, axis=1)/2.
        X = np.hstack((X[:, [0]] - dX[:, [0]],
                       X[:, :-1] + dX,
                       X[:, [-1]] + dX[:, [-1]]))
    return X

x = np.array([1, 2, 0])
y = np.array([1, 2])

X, Y = np.meshgrid(x, y)

_interp_grid(X)
```

**Actual outcome**

Results in
```python
array([[ 0.5,  1.5,  1. , -1. ],
       [ 0.5,  1.5,  1. , -1. ]])
```

While having cooridinates `[1, 2, 0]` may seem just as wrong as coordinates `[0.5, 1.5, 1, -1]` the latter can cause some real-world problems in conjunction with cartopy (scitools/cartopy#1638):

![shading_rasm_both](https://user-images.githubusercontent.com/10194086/90925277-3ab84600-e3f1-11ea-8aa5-8debfb465495.png)

**Expected outcome**

* A warning for non-monotonic data
* A convenient option to restore the old behaviour, e.g. `shading="drop"`

cc @greglucas

**Matplotlib version**

  * Operating system: Linux Mint
  * Matplotlib version: '3.3.1'
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.8
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by jklymak** [comment]
So, it seems to me that the problem with my new algorithm and cartopy wraps is that the wrap masking logic is just not kicking in for small enough wraps.  By interpolating the wraps are made smaller, i.e. 1/3 of the panel rather than 1/2, and they aren't being masked.

**Comment by greglucas** [comment]
I don't think that is quite it. Even if we identified those cells as masked they would have a bad point in them and the pcolor algorithm would use that bad point to draw the polygons. I think that we need to address this on the Cartopy side because there are two competing data wraps (179 -> 181 and 359 -> 1). The way I think about it now is that Matplotlib is trying to be nice for users and make the edges that `pcolormesh` really expects. Cartopy in this case can be a user and calculate the edges before calling pcolormesh and negate the need to use the "auto" shading.

**Comment by jklymak** [comment]
That would be ideal - or at worst case downstream libraries can drop the last row and column if they can't come up with a better way of specifying the corners.

OTOH, I think 1) *matplotlib* could still consider a screen space midpoint rather than a data space one.  2) matplotlib could warn if whatever interpolation we decide on is non-monotonic in any rows for X and any columns for Y

**Comment by jklymak** [comment]
I believe this is closed by #18398....

**Comment by mathause** [comment]
thanks everyone!
